File: lohas-papers-backend/app/external/pubmed.py
import asyncio
import logging
import xml.etree.ElementTree as ET

import httpx

from app.config import get_settings
from app.models.schemas import UnifiedPaper

logger = logging.getLogger(__name__)

ESEARCH_URL = "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi"
EFETCH_URL = "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi"
TIMEOUT = 10.0


async def search_papers(
    query: str,
    *,
    limit: int = 20,
    year_from: int | None = None,
    year_to: int | None = None,
) -> list[UnifiedPaper]:
    """Search PubMed for papers matching the query."""
    settings = get_settings()

    # Add date filter to query if specified
    date_filter = ""
    if year_from:
        date_filter += f" AND {year_from}[PDAT]"
    if year_to:
        date_filter += f" AND {year_to}[PDAT]"

    search_params: dict = {
        "db": "pubmed",
        "term": query + date_filter,
        "retmax": limit,
        "sort": "relevance",
        "retmode": "json",
    }
    if settings.pubmed_api_key:
        search_params["api_key"] = settings.pubmed_api_key

    xml_text = None
    max_retries = 3
    for attempt in range(max_retries):
        try:
            async with httpx.AsyncClient(timeout=TIMEOUT) as client:
                # Step 1: ESearch to get PMIDs
                resp = await client.get(ESEARCH_URL, params=search_params)
                resp.raise_for_status()
                search_data = resp.json()

                id_list = search_data.get("esearchresult", {}).get("idlist", [])
                if not id_list:
                    return []

                # Step 2: EFetch to get full records
                fetch_params: dict = {
                    "db": "pubmed",
                    "id": ",".join(id_list),
                    "retmode": "xml",
                }
                if settings.pubmed_api_key:
                    fetch_params["api_key"] = settings.pubmed_api_key

                resp = await client.get(EFETCH_URL, params=fetch_params)
                resp.raise_for_status()
                xml_text = resp.text
                break
        except httpx.HTTPStatusError as e:
            if e.response.status_code == 429 and attempt < max_retries - 1:
                wait = 2 ** attempt + 1
                logger.info("PubMed 429, retrying in %ds: %s", wait, query)
                await asyncio.sleep(wait)
                continue
            logger.warning("PubMed HTTP error %s: %s", e.response.status_code, query)
            return []
        except httpx.TimeoutException:
            logger.warning("PubMed timeout for query: %s", query)
            return []
        except Exception:
            logger.exception("PubMed unexpected error for query: %s", query)
            return []

    if xml_text is None:
        return []

    return _parse_pubmed_xml(xml_text)


def _parse_pubmed_xml(xml_text: str) -> list[UnifiedPaper]:
    """Parse PubMed XML response into UnifiedPaper objects."""
    papers: list[UnifiedPaper] = []

    try:
        root = ET.fromstring(xml_text)
    except ET.ParseError:
        logger.error("Failed to parse PubMed XML")
        return []

    for article_elem in root.findall(".//PubmedArticle"):
        try:
            paper = _parse_single_article(article_elem)
            if paper:
                papers.append(paper)
        except Exception:
            logger.warning("Failed to parse PubMed article", exc_info=True)
            continue

    logger.info("PubMed returned %d papers", len(papers))
    return papers


def _parse_single_article(article_elem: ET.Element) -> UnifiedPaper | None:
    """Parse a single PubmedArticle XML element."""
    medline = article_elem.find(".//MedlineCitation")
    if medline is None:
        return None

    # PMID
    pmid_elem = medline.find("PMID")
    pmid = pmid_elem.text if pmid_elem is not None else None
    if not pmid:
        return None

    article = medline.find("Article")
    if article is None:
        return None

    # Title
    title_elem = article.find("ArticleTitle")
    title = title_elem.text if title_elem is not None else ""

    # Abstract
    abstract_parts = []
    abstract_elem = article.find("Abstract")
    if abstract_elem is not None:
        for abs_text in abstract_elem.findall("AbstractText"):
            label = abs_text.get("Label", "")
            text = abs_text.text or ""
            if label:
                abstract_parts.append(f"{label}: {text}")
            else:
                abstract_parts.append(text)
    abstract = " ".join(abstract_parts) if abstract_parts else None

    # Authors
    authors = []
    author_list = article.find("AuthorList")
    if author_list is not None:
        for author in author_list.findall("Author"):
            last = author.findtext("LastName", "")
            first = author.findtext("ForeName", "")
            if last:
                authors.append(f"{first} {last}".strip())

    # Journal
    journal_elem = article.find("Journal/Title")
    journal = journal_elem.text if journal_elem is not None else None

    # Year
    year = None
    pub_date = article.find("Journal/JournalIssue/PubDate")
    if pub_date is not None:
        year_elem = pub_date.find("Year")
        if year_elem is not None and year_elem.text:
            try:
                year = int(year_elem.text)
            except ValueError:
                pass

    # DOI
    doi = None
    for eid in article.findall("ELocationID"):
        if eid.get("EIdType") == "doi":
            doi = eid.text
            break

    return UnifiedPaper(
        id=f"pmid:{pmid}",
        title=title,
        authors=authors,
        journal=journal,
        year=year,
        doi=doi,
        pmid=pmid,
        citation_count=0,  # PubMed doesn't provide citation counts directly
        is_open_access=False,  # Would need PMC check
        pdf_url=None,
        abstract=abstract,
        source="pubmed",
    )

File: lohas-papers-web/src/lib/backend/pubmed.ts
import type { UnifiedPaper } from "./types";

const ESEARCH_URL = "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi";
const EFETCH_URL = "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi";
const TIMEOUT_MS = 10_000;
const TOOL_NAME = "lohas-papers";

let nextPubMedRequestAt = 0;

export async function searchPapers(
  query: string,
  options?: {
    limit?: number;
    yearFrom?: number | null;
    yearTo?: number | null;
  },
): Promise<UnifiedPaper[]> {
  const { limit = 20, yearFrom, yearTo } = options ?? {};
  const pubmedApiKey = process.env.PUBMED_API_KEY || "";
  const pubmedEmail = process.env.PUBMED_EMAIL || "";

  // Add date filter to query if specified
  let dateFilter = "";
  if (yearFrom) dateFilter += ` AND ${yearFrom}[PDAT]`;
  if (yearTo) dateFilter += ` AND ${yearTo}[PDAT]`;

  const searchParams = new URLSearchParams({
    db: "pubmed",
    term: query + dateFilter,
    retmax: String(limit),
    sort: "relevance",
    retmode: "json",
    tool: TOOL_NAME,
  });
  if (pubmedApiKey) searchParams.set("api_key", pubmedApiKey);
  if (pubmedEmail) searchParams.set("email", pubmedEmail);

  // Step 1: ESearch to get PMIDs
  const searchResp = await fetchPubMedWithRetries(
    `${ESEARCH_URL}?${searchParams}`,
    query,
    "PubMed ESearch",
    Boolean(pubmedApiKey),
  );
  if (!searchResp) return [];

  const searchData = await searchResp.json();
  const idList: string[] = searchData?.esearchresult?.idlist ?? [];
  if (idList.length === 0) return [];

  // Step 2: EFetch to get full records
  const fetchParams = new URLSearchParams({
    db: "pubmed",
    id: idList.join(","),
    retmode: "xml",
    tool: TOOL_NAME,
  });
  if (pubmedApiKey) fetchParams.set("api_key", pubmedApiKey);
  if (pubmedEmail) fetchParams.set("email", pubmedEmail);

  const fetchResp = await fetchPubMedWithRetries(
    `${EFETCH_URL}?${fetchParams}`,
    query,
    "PubMed EFetch",
    Boolean(pubmedApiKey),
  );
  if (!fetchResp) return [];

  const xmlText = await fetchResp.text();
  if (!xmlText) {
    return [];
  }

  return parsePubmedXml(xmlText);
}

// ── XML Parsing (lightweight, no external dependency) ──

function parsePubmedXml(xmlText: string): UnifiedPaper[] {
  const papers: UnifiedPaper[] = [];

  // Split on PubmedArticle tags
  const articleRegex = /<PubmedArticle>([\s\S]*?)<\/PubmedArticle>/g;
  let match: RegExpExecArray | null;

  while ((match = articleRegex.exec(xmlText)) !== null) {
    try {
      const paper = parseSingleArticle(match[1]);
      if (paper) papers.push(paper);
    } catch {
      console.warn("Failed to parse PubMed article");
    }
  }

  console.info(`PubMed returned ${papers.length} papers`);
  return papers;
}

function parseSingleArticle(articleXml: string): UnifiedPaper | null {
  // PMID
  const pmid = extractTag(articleXml, "PMID");
  if (!pmid) return null;

  // Title
  const title = extractTag(articleXml, "ArticleTitle") || "";

  // Abstract
  const abstractParts: string[] = [];
  const abstractTextRegex =
    /<AbstractText(?:\s+Label="([^"]*)")?[^>]*>([\s\S]*?)<\/AbstractText>/g;
  let absMatch: RegExpExecArray | null;
  while ((absMatch = abstractTextRegex.exec(articleXml)) !== null) {
    const label = absMatch[1] || "";
    const text = stripTags(absMatch[2] || "");
    abstractParts.push(label ? `${label}: ${text}` : text);
  }
  const abstract = abstractParts.length > 0 ? abstractParts.join(" ") : undefined;

  // Authors
  const authors: string[] = [];
  const authorRegex = /<Author[^>]*>([\s\S]*?)<\/Author>/g;
  let authorMatch: RegExpExecArray | null;
  while ((authorMatch = authorRegex.exec(articleXml)) !== null) {
    const last = extractTag(authorMatch[1], "LastName") || "";
    const first = extractTag(authorMatch[1], "ForeName") || "";
    if (last) authors.push(`${first} ${last}`.trim());
  }

  // Journal
  const journal = extractTag(articleXml, "Title");

  // Year
  let year: number | undefined;
  const yearMatch = /<PubDate>([\s\S]*?)<\/PubDate>/.exec(articleXml);
  if (yearMatch) {
    const yearStr = extractTag(yearMatch[1], "Year");
    if (yearStr) {
      const parsed = parseInt(yearStr, 10);
      if (!isNaN(parsed)) year = parsed;
    }
  }

  // DOI
  let doi: string | undefined;
  const doiRegex = /<ELocationID\s+EIdType="doi"[^>]*>([\s\S]*?)<\/ELocationID>/;
  const doiMatch = doiRegex.exec(articleXml);
  if (doiMatch) doi = stripTags(doiMatch[1]).trim();

  return {
    id: `pmid:${pmid}`,
    title,
    authors,
    journal: journal || null,
    year: year ?? null,
    doi: doi || null,
    pmid,
    citation_count: 0, // PubMed doesn't provide citation counts directly
    is_open_access: false, // Would need PMC check
    pdf_url: null,
    abstract: abstract || null,
    source: "pubmed",
  };
}

function extractTag(xml: string, tag: string): string | null {
  const regex = new RegExp(`<${tag}[^>]*>([\\s\\S]*?)<\\/${tag}>`);
  const match = regex.exec(xml);
  return match ? stripTags(match[1]).trim() : null;
}

function stripTags(html: string): string {
  return html.replace(/<[^>]*>/g, "");
}

async function fetchPubMedWithRetries(
  url: string,
  query: string,
  label: string,
  hasApiKey: boolean,
): Promise<Response | null> {
  const maxRetries = 3;

  for (let attempt = 0; attempt < maxRetries; attempt++) {
    try {
      await waitForPubMedSlot(hasApiKey);
      const resp = await fetch(url, {
        signal: AbortSignal.timeout(TIMEOUT_MS),
      });

      if (resp.ok) return resp;

      if (resp.status === 429 && attempt < maxRetries - 1) {
        const waitMs = getRetryDelayMs(resp.headers.get("retry-after"), attempt);
        console.info(
          `${label} 429, retrying in ${Math.round(waitMs / 1000)}s: ${query}`,
        );
        await sleep(waitMs);
        continue;
      }

      console.warn(`${label} HTTP error ${resp.status}: ${query}`);
      return null;
    } catch (err) {
      if (err instanceof DOMException && err.name === "TimeoutError") {
        console.warn(`${label} timeout for query: ${query}`);
        return null;
      }
      console.error(`${label} unexpected error for query: ${query}`, err);
      return null;
    }
  }

  return null;
}

async function waitForPubMedSlot(hasApiKey: boolean): Promise<void> {
  const minIntervalMs = hasApiKey ? 120 : 450;
  const now = Date.now();
  const waitMs = Math.max(0, nextPubMedRequestAt - now);
  nextPubMedRequestAt = Math.max(now, nextPubMedRequestAt) + minIntervalMs;
  if (waitMs > 0) {
    await sleep(waitMs);
  }
}

function getRetryDelayMs(retryAfter: string | null, attempt: number): number {
  if (retryAfter) {
    const seconds = Number(retryAfter);
    if (Number.isFinite(seconds) && seconds > 0) {
      return seconds * 1000;
    }

    const retryAt = Date.parse(retryAfter);
    if (Number.isFinite(retryAt)) {
      return Math.max(1000, retryAt - Date.now());
    }
  }

  return (attempt + 2) * 1500;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
